be/src/storage/rowset/segment_iterator.cpp

```diff
@@ -682,6 +699,25 @@ Status SegmentIterator::_do_get_next(Chunk* result, vector<rowid_t>* rowid) {
         DCHECK_EQ(chunk_start, chunk->num_rows());
     }
 
+    // If _read_chunk contains more rows than its capacity, we save the overflow rows in _overflow_read_chunk.
+    if (chunk_start > chunk_capacity) {
+        if (_overflow_read_chunk == nullptr) {
+            _overflow_read_chunk = chunk->clone_empty();
+            if (rowid != nullptr) {
+                _overflow_read_chunk_rowids.reserve(_opts.chunk_size);
+            }
+        }
+        DCHECK(_overflow_read_chunk->is_empty());
+        _overflow_read_chunk->append(*chunk, chunk_capacity, chunk_start - chunk_capacity);
+        if (rowid != nullptr) {
+            DCHECK(_overflow_read_chunk_rowids.empty());
+            _overflow_read_chunk_rowids.insert(_overflow_read_chunk_rowids.end(), rowid->begin() + chunk_capacity,
+                                               rowid->end());
+            rowid->resize(chunk_capacity);
+        }
+        chunk->set_num_rows(chunk_capacity);
```

Comment:
chunk->set_num_rows() `resize` data but not `shrink_to_fit`, it make all chunk actual memory consumption larger than 4096 rows?

Reply:
Each time `do_get_next`, `chunk` is likely to contain more than chunk_capacity rows. I think `shrink_to_fit` here is not a good choice? If we do so, the frequent shrink and grow will have some overhead?

Reply:
I don't mean to call shrink_to_fit, but is this way of consuming extra memory worth it?

Reply:
previous implementation _read() size is `chunk_capacity - chunk_start`, Doesn't seem to exceed chunk capacity

Reply:
When chunk_start equals to `chunk_capacity - 1`, _read only scan one more line each time. If the row satisfied the predicate is last row in the worst case. Each time `_read` can only retrieve one line, the overhead of the function call will be very large.